### PR TITLE
Virtualize trials list

### PR DESCRIPTION
**Pull request by y0z**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## What does this implement/fix? Explain your changes.

It is known that the Optuna dashboard seriously slows down to show the list of trials when the number of trials is too large (e.g., N > 100,000). This PR virtualizes the list of trials to accelerate rendering in the `dashboard/studies/:id/trials` page.
- Note that this PR introduces a dependency on [`react-virtual`](https://tanstack.com/virtual/latest/docs/introduction).

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@keisuke-umezawa Could you review this PR?

File: optuna_dashboard/ts/components/TrialList.tsx [hunk 1]
```diff
@@ -339,6 +341,14 @@ export const TrialList: FC<{ studyDetail: StudyDetail | null }> = ({
       (state) => allTrials.filter((t) => t.state === state).length
     )
   }, [studyDetail?.trials])
+  const listParentRef = React.useRef(null)
+
+  const rowVirtualizer = useVirtualizer({
+    count: trials.length,
+    getScrollElement: () => listParentRef.current,
+    estimateSize: () => 75,
+    overscan: 10,
```
Comment (keisuke-umezawa) [inline optuna_dashboard/ts/components/TrialList.tsx:350]: [question] How did you decide those numbers?

File: optuna_dashboard/ts/components/TrialList.tsx [hunk 2]
```diff
@@ -412,65 +423,91 @@ export const TrialList: FC<{ studyDetail: StudyDetail | null }> = ({
...
-                  }
+          <Box
+            sx={{
+              width: "100%",
+              height: `${rowVirtualizer.getTotalSize()}px`,
+              position: "relative",
+            }}
+          >
+            {rowVirtualizer.getVirtualItems().map((virtualItem) => {
+              console.log(virtualItem.index, virtualItem.start, virtualItem.end)
```
Comment (keisuke-umezawa) [inline optuna_dashboard/ts/components/TrialList.tsx:434]: [nits] Is it for debugging?

**Review by keisuke-umezawa — CHANGES_REQUESTED** [review]
It seems to work as expected, but I can not run it with large trials with my local env. So, could you do test or QA with large trials?

Also, I found a minor difference of UI. Could you fix it?
スクリーンショット 2024-03-02 15 24 53

File: optuna_dashboard/ts/components/TrialList.tsx (hunk 2 above)
Comment (y0z) [reply optuna_dashboard/ts/components/TrialList.tsx]: Thank you for your comment.
I deleted this since it is for debugging.

**Comment by y0z** [comment]
Thank you for your comment.
I've fixed this.

> Also, I found a minor difference of UI. Could you fix it?
image

File: optuna_dashboard/ts/components/TrialList.tsx (hunk 1 above)
Comment (y0z) [reply optuna_dashboard/ts/components/TrialList.tsx:350]: I've changed the value for estimateSize to 73.31 based on the actual height.
image

On the other hand, the value for overscan was determined heuristically and I don't have a strong opinion.
Do you have any ideas on how to decide this value?

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! It also worked in my env. Could you resolve conflicts?

**Comment by y0z** [comment]
It seems that CI is broken.
https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810

```
Run npm run lint

> optuna-dashboard@0.0.[1](https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810#step:5:1) lint
> npm run lint:eslint && npm run lint:biome


> optuna-dashboard@0.0.1 lint:eslint
> eslint . --ext .ts,.tsx --max-warnings 0


> optuna-dashboard@0.0.1 lint:biome
> biome format . && biome ci standalone_app vscode

node:internal/modules/cjs/loader:1031
  throw err;
  ^

Error: Cannot find module '@biomejs/cli-linux-x6[4](https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810#step:5:5)/biome'
Require stack:
- /home/runner/work/optuna-dashboard/optuna-dashboard/node_modules/@biomejs/biome/bin/biome
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1028:1[5](https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810#step:5:6))
    at Function.resolve (node:internal/modules/cjs/helpers:125:19)
    at Object.<anonymous> (/home/runner/work/optuna-dashboard/optuna-dashboard/node_modules/@biomejs/biome/bin/biome:51:11)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:107[6](https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810#step:5:7):32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:4[7](https://github.com/optuna/optuna-dashboard/actions/runs/8320354360/job/22764945809?pr=810#step:5:8) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/runner/work/optuna-dashboard/optuna-dashboard/node_modules/@biomejs/biome/bin/biome'
  ]
}
Error: Process completed with exit code 1.
```

**Comment by y0z** [comment]
> LGTM! It also worked in my env. Could you resolve conflicts?

The conflicts have been resolved :)

**Review by nabenabe0928 — COMMENTED** [review]
I will also quickly take a look at UI changes.

File: optuna_dashboard/ts/components/TrialList.tsx [hunk 3]
```diff
@@ -412,65 +423,91 @@ export const TrialList: FC<{ studyDetail: StudyDetail | null }> = ({
...
+                <ListItem
+                  key={trial.trial_id}
                   sx={{
                     display: "flex",
                     flexDirection: "column",
                     alignItems: "flex-start",
+                    position: "absolute",
+                    top: 0,
+                    left: 0,
+                    transform: `translateY(${virtualItem.start}px)`,
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/TrialList.tsx:445]: changed

File: optuna_dashboard/ts/components/TrialList.tsx [hunk 4]
```diff
@@ -412,65 +423,91 @@ export const TrialList: FC<{ studyDetail: StudyDetail | null }> = ({
...
-                  selected={
-                    selected.findIndex((t) => t.number === trial.number) !== -1
-                  }
+          <Box
+            sx={{
+              width: "100%",
+              height: `${rowVirtualizer.getTotalSize()}px`,
+              position: "relative",
+            }}
+          >
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/TrialList.tsx:432]: changed

File: optuna_dashboard/ts/components/TrialList.tsx [hunk 5]
```diff
@@ -348,13 +358,14 @@ export const TrialList: FC<{ studyDetail: StudyDetail | null }> = ({
...
       <Box
+        ref={listParentRef}
         sx={{
           minWidth: trialListWidth,
           overflow: "auto",
           height: `calc(100vh - ${theme.spacing(8)})`,
         }}
       >
-        <List>
+        <List sx={{ position: "relative" }}>
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/TrialList.tsx:368]: changed

**Review by nabenabe0928 — COMMENTED** [review]
I just left memos for me to mark the changed parts.

**Review by porink0424 — APPROVED** [review]
LGTM👍

**Pull request merged**